packages/hub/src/lib/parse-safetensors.ts

```diff
@@ -0,0 +1,136 @@
+	dtype: Dtype;
+	shape: number[];
+	data_offsets: [number, number];
+}
+
+type FileHeader = Record<TensorName, TensorInfo> & {
+	__metadata__: Record<string, string>;
+};
+
+interface IndexJson {
+	dtype?: string;
+	/// ^there's sometimes a dtype but it looks inconsistent.
+	metadata?: Record<string, string>;
+	/// ^ why the naming inconsistency?
+	weight_map: Record<TensorName, FileName>;
+}
+
+type ShardedHeaders = Record<FileName, FileHeader>;
+
+type ParseFromRepo =
+	| {
+			sharded: false;
+			header: FileHeader;
+	  }
+	| {
+			sharded: true;
+			index: IndexJson;
+			headers: ShardedHeaders;
+	  };
+
+async function parseSingleFile(url: URL): Promise<FileHeader> {
+	const bufLengthOfHeaderLE = await (
+		await fetch(url, {
+			headers: {
+				Range: "bytes=0-7",
+			},
+		})
+	).arrayBuffer();
+	const lengthOfHeader = new DataView(bufLengthOfHeaderLE).getBigUint64(0, true);
+	/// ^little-endian
+	const header: FileHeader = await (
+		await fetch(url, {
+			headers: {
+				Range: `bytes=8-${7 + Number(lengthOfHeader)}`,
+			},
+		})
+	).json();
+	/// no validation for now, we assume it's a valid FileHeader.
+	return header;
+}
+
+async function parseShardedIndex(url: URL): Promise<{ index: IndexJson; headers: ShardedHeaders }> {
+	const index: IndexJson = await (await fetch(url)).json();
+	/// no validation for now, we assume it's a valid IndexJson.
+
+	const shardedMap: ShardedHeaders = {};
+	const filenames = [...new Set(Object.values(index.weight_map))];
+	await Promise.all(
+		filenames.map(async (filename) => {
+			const singleUrl = new URL(url.toString().replace(INDEX_FILE, filename));
+			shardedMap[filename] = await parseSingleFile(singleUrl);
+		})
+	);
+	return { index, headers: shardedMap };
+}
+
+async function doesFileExistOnHub(url: URL): Promise<boolean> {
+	const res = await fetch(url, {
+		method: "HEAD",
+		redirect: "manual",
+		/// ^do not follow redirects to save some time
+	});
+	return res.status >= 200 && res.status < 400;
+}
+
+export async function parseSafetensorsFromModelRepo(id: string): Promise<ParseFromRepo> {
+	const singleUrl = new URL(`https://huggingface.co/${id}/resolve/main/${SINGLE_FILE}`);
+	const indexUrl = new URL(`https://huggingface.co/${id}/resolve/main/${INDEX_FILE}`);
+	if (await doesFileExistOnHub(singleUrl)) {
+		return {
+			sharded: false,
+			header: await parseSingleFile(singleUrl),
+		};
+	} else if (await doesFileExistOnHub(indexUrl)) {
+		return {
+			sharded: true,
+			...(await parseShardedIndex(indexUrl)),
+		};
+	} else {
+		throw new Error("model id does not seem to contain safetensors weights");
+	}
+}
+
+function computeNumOfParamsByDtypeSingleFile(header: FileHeader): Counter<Dtype> {
+	const n = new Counter<Dtype>();
+	for (const [k, v] of Object.entries(header)) {
+		if (k === "__metadata__") {
+			continue;
+		}
+		if ((v as TensorInfo).shape.length === 0) {
+			continue;
+		}
+		n.incr(
+			(v as TensorInfo).dtype,
+			(v as TensorInfo).shape.reduce((a, b) => a * b)
+		);
```

Comment:
```suggestion
		n.set(v.dtype, (n.get(v.dtype) ?? 0) + sum(v.shape.map((a, b) => a * b));
```

We can add a `sum` utils. I don't think we need to add a whole `Counter` class

Reply:
Are you sure? Given i'm also using .dict and .total, I'm wondering if it's going to be a bit verbose/error-prone to not use it

Reply:
i've slimmed it down a litte bit in 2cc913d7097ce50fda339fba1f6ad74a1587ebb1

Reply:
i can still remove it if you think it's better, though

Reply:
(it's not a sum BTW, it's really the dot product of everything a * b  * c * ...)

Reply:
(yes noticed afterwards ^^)

Reply:
3757fc3598049bf3eba451e5e363a958f00f25b3